File: frontend/src/types/comprobante.ts
/**
 * Tipos TypeScript para Comprobantes
 */

export interface ItemComprobante {
  id?: number
  codigo?: string
  descripcion: string
  cantidad: number
  unidad: string
  precio_unitario: number
  descuento_porcentaje: number
  iva_porcentaje: number
  subtotal?: number
  orden: number
  comprobante_id?: number
}

export interface EmitirComprobanteRequest {
  empresa_id: number
  punto_venta_id: number
  tipo_comprobante: number
  concepto: number
  
  // Cliente
  cliente_id?: number
  tipo_documento: number
  numero_documento: string
  razon_social: string
  condicion_iva: string
  domicilio?: string
  
  // Items
  items: ItemComprobante[]
  
  // Servicios
  fecha_servicio_desde?: string
  fecha_servicio_hasta?: string
  fecha_vto_pago?: string
  
  // Opcional
  observaciones?: string
  moneda: string
  cotizacion: number
}

export interface EmitirComprobanteResponse {
  exito: boolean
  comprobante_id?: number
  tipo_comprobante: number
  punto_venta: number
  numero: number
  fecha: string
  cae?: string
  cae_vencimiento?: string
  total: number
  mensaje: string
  errores: string[]
}

export interface Comprobante {
  id: number
  tipo_comprobante: number
  numero: number
  fecha_emision: string
  fecha_vencimiento?: string
  subtotal: number
  descuento: number
  iva_21: number
  iva_10_5: number
  iva_27: number
  otros_impuestos: number
  total: number
  cae?: string
  cae_vencimiento?: string
  estado: string
  moneda: string
  cotizacion: number
  observaciones?: string
  empresa_id: number
  punto_venta_id: number
  cliente_id: number
}

export interface ComprobanteDetalle extends Comprobante {
  items: ItemComprobante[]
  cliente_nombre?: string
  cliente_cuit?: string
  punto_venta_numero?: number
}

export interface ComprobanteListItem {
  id: number
  tipo_comprobante: number
  numero: number
  fecha_emision: string
  total: number
  estado: string
  cae?: string
  cliente_nombre: string
  cliente_documento: string
  punto_venta_numero: number
}

export interface PaginatedComprobantesResponse {
  items: ComprobanteListItem[]
  total: number
  page: number
  per_page: number
  pages: number
}

export interface ProximoNumeroResponse {
  punto_venta: number
  tipo_comprobante: number
  proximo_numero: number
}

// Tipos de comprobante
export const TIPOS_COMPROBANTE = {
  FACTURA_A: 1,
  NOTA_DEBITO_A: 2,
  NOTA_CREDITO_A: 3,
  FACTURA_B: 6,
  NOTA_DEBITO_B: 7,
  NOTA_CREDITO_B: 8,
  FACTURA_C: 11,
  NOTA_DEBITO_C: 12,
  NOTA_CREDITO_C: 13,
} as const

export const TIPOS_COMPROBANTE_NOMBRES: Record<number, string> = {
  1: 'Factura A',
  2: 'Nota de Débito A',
  3: 'Nota de Crédito A',
  6: 'Factura B',
  7: 'Nota de Débito B',
  8: 'Nota de Crédito B',
  11: 'Factura C',
  12: 'Nota de Débito C',
  13: 'Nota de Crédito C',
}

// Tipos de concepto
export const TIPOS_CONCEPTO = {
  PRODUCTOS: 1,
  SERVICIOS: 2,
  PRODUCTOS_Y_SERVICIOS: 3,
} as const

export const TIPOS_CONCEPTO_NOMBRES: Record<number, string> = {
  1: 'Productos',
  2: 'Servicios',
  3: 'Productos y Servicios',
}

// Tipos de documento
export const TIPOS_DOCUMENTO = {
  CUIT: 80,
  CUIL: 86,
  CDI: 87,
  LE: 89,
  LC: 90,
  CI_EXTRANJERA: 91,
  EN_TRAMITE: 92,
  ACTA_NACIMIENTO: 93,
  CI_BS_AS_RNP: 95,
  DNI: 96,
  PASAPORTE: 94,
  CI_POLICIA_FEDERAL: 0,
  CI_BUENOS_AIRES: 1,
  CI_CATAMARCA: 2,
  CI_CORDOBA: 3,
  CI_CORRIENTES: 4,
  CI_ENTRE_RIOS: 5,
  CI_JUJUY: 6,
  CI_MENDOZA: 7,
  CI_LA_RIOJA: 8,
  CI_SALTA: 9,
  CI_SAN_JUAN: 10,
  CI_SAN_LUIS: 11,
  CI_SANTA_FE: 12,
  CI_SANTIAGO_DEL_ESTERO: 13,
  CI_TUCUMAN: 14,
  CI_CHACO: 16,
  CI_CHUBUT: 17,
  CI_FORMOSA: 18,
  CI_MISIONES: 19,
  CI_NEUQUEN: 20,
  CI_LA_PAMPA: 21,
  CI_RIO_NEGRO: 22,
  CI_SANTA_CRUZ: 23,
  CI_TIERRA_DEL_FUEGO: 24,
} as const

export const TIPOS_DOCUMENTO_NOMBRES: Record<number, string> = {
  80: 'CUIT',
  86: 'CUIL',
  96: 'DNI',
  94: 'Pasaporte',
  99: 'Consumidor Final',
}

// Alícuotas de IVA
export const ALICUOTAS_IVA = [
  { value: 0, label: 'Exento / 0%', porcentaje: 0 },
  { value: 10.5, label: '10.5%', porcentaje: 10.5 },
  { value: 21, label: '21%', porcentaje: 21 },
  { value: 27, label: '27%', porcentaje: 27 },
]

// Condiciones IVA
export const CONDICIONES_IVA = [
  'Responsable Inscripto',
  'Monotributo',
  'Exento',
  'Consumidor Final',
  'Responsable No Inscripto',
]

// Estados de comprobante
export const ESTADOS_COMPROBANTE = {
  BORRADOR: 'borrador',
  PENDIENTE: 'pendiente',
  AUTORIZADO: 'autorizado',
  RECHAZADO: 'rechazado',
  ANULADO: 'anulado',
} as const

export const ESTADOS_COMPROBANTE_NOMBRES: Record<string, string> = {
  borrador: 'Borrador',
  pendiente: 'Pendiente',
  autorizado: 'Autorizado',
  rechazado: 'Rechazado',
  anulado: 'Anulado',
}

File: backend/app/schemas/comprobante.py
"""Schemas de Comprobante - Facturas, Notas de Crédito y Débito."""

from datetime import date
from decimal import Decimal
from typing import List, Optional
from pydantic import BaseModel, Field, field_validator

# ==================== Items ====================


class ItemComprobanteBase(BaseModel):
    """Schema base de Item de Comprobante."""

    codigo: Optional[str] = None
    descripcion: str = Field(..., min_length=1, max_length=500)
    cantidad: Decimal = Field(..., gt=0)
    unidad: str = Field(default="unidades", max_length=50)
    precio_unitario: Decimal = Field(..., ge=0)
    descuento_porcentaje: Decimal = Field(default=Decimal("0"), ge=0, le=100)
    iva_porcentaje: Decimal = Field(..., ge=0)  # 21, 10.5, 27, 0
    orden: int = Field(default=0, ge=0)


class ItemComprobanteCreate(ItemComprobanteBase):
    """Schema para crear Item de Comprobante."""

    pass


class ItemComprobanteResponse(ItemComprobanteBase):
    """Schema de respuesta de Item de Comprobante."""

    id: int
    subtotal: Decimal
    comprobante_id: int

    class Config:
        from_attributes = True


# ==================== Comprobante ====================


class ComprobanteBase(BaseModel):
    """Schema base de Comprobante."""

    tipo_comprobante: int = Field(..., ge=1, le=999)
    observaciones: Optional[str] = Field(None, max_length=500)
    moneda: str = Field(default="PES", max_length=3)
    cotizacion: Decimal = Field(default=Decimal("1"), gt=0)


class EmitirComprobanteRequest(ComprobanteBase):
    """
    Request para emitir un comprobante.

    Incluye todos los datos necesarios para generar
    el comprobante y solicitar el CAE a ARCA.
    """

    empresa_id: int
    punto_venta_id: int
    concepto: int = Field(default=1, ge=1, le=3)  # 1=Productos, 2=Servicios, 3=Ambos

    # Cliente/Receptor
    cliente_id: Optional[int] = None  # Si es cliente guardado
    tipo_documento: int = Field(..., ge=1)  # 80=CUIT, 96=DNI, etc.
    numero_documento: str = Field(..., min_length=1, max_length=20)
    razon_social: str = Field(..., min_length=1, max_length=200)
    condicion_iva: str = Field(..., max_length=50)
    domicilio: Optional[str] = Field(None, max_length=200)

    # Items
    items: List[ItemComprobanteCreate] = Field(..., min_length=1)

    # Servicios (si concepto = 2 o 3)
    fecha_servicio_desde: Optional[date] = None
    fecha_servicio_hasta: Optional[date] = None
    fecha_vto_pago: Optional[date] = None

    @field_validator("items")
    @classmethod
    def validate_items(cls, v):
        if not v or len(v) == 0:
            raise ValueError("Debe incluir al menos un ítem")
        return v

    @field_validator("fecha_servicio_hasta")
    @classmethod
    def validate_fecha_servicio(cls, v, info):
        if v and info.data.get("fecha_servicio_desde"):
            if v < info.data["fecha_servicio_desde"]:
                raise ValueError("La fecha hasta no puede ser menor a la fecha desde")
        return v


class EmitirComprobanteResponse(BaseModel):
    """Respuesta al emitir un comprobante."""

    exito: bool
    comprobante_id: Optional[int] = None
    tipo_comprobante: int
    punto_venta: int
    numero: int
    fecha: date
    cae: Optional[str] = None
    cae_vencimiento: Optional[date] = None
    total: Decimal
    mensaje: str
    errores: List[str] = Field(default_factory=list)


class ComprobanteResponse(ComprobanteBase):
    """Schema de respuesta de Comprobante."""

    id: int
    numero: int
    fecha_emision: date
    fecha_vencimiento: Optional[date] = None
    subtotal: Decimal
    descuento: Decimal
    iva_21: Decimal
    iva_10_5: Decimal
    iva_27: Decimal
    otros_impuestos: Decimal
    total: Decimal
    cae: Optional[str] = None
    cae_vencimiento: Optional[date] = None
    estado: str

    # Relaciones
    empresa_id: int
    punto_venta_id: int
    cliente_id: int

    class Config:
        from_attributes = True


class ComprobanteDetalleResponse(ComprobanteResponse):
    """Schema de respuesta de Comprobante con detalles completos."""

    items: List[ItemComprobanteResponse] = Field(default_factory=list)

    # Datos del cliente
    cliente_nombre: Optional[str] = None
    cliente_cuit: Optional[str] = None

    # Datos del punto de venta
    punto_venta_numero: Optional[int] = None

    class Config:
        from_attributes = True


class ComprobanteListResponse(BaseModel):
    """Schema para listado de comprobantes."""

    id: int
    tipo_comprobante: int
    numero: int
    fecha_emision: date
    total: Decimal
    estado: str
    cae: Optional[str] = None

    # Datos básicos del cliente
    cliente_nombre: str
    cliente_documento: str

    # Datos del punto de venta
    punto_venta_numero: int

    class Config:
        from_attributes = True


class PaginatedComprobantesResponse(BaseModel):
    """Respuesta paginada de comprobantes."""

    items: List[ComprobanteListResponse]
    total: int
    page: int
    per_page: int
    pages: int


class ProximoNumeroResponse(BaseModel):
    """Respuesta con el próximo número de comprobante."""

    punto_venta: int
    tipo_comprobante: int
    proximo_numero: int
